### PR TITLE
Simplify element table construction

### PR DESCRIPTION
**Pull request by chrisrichardson**
Removes some 400 lines of code, most of which was involved with creating "unique" element tables.

After removing several layers, the code is now simple, and just creates the tables for each modified terminal, checking as it goes along that they are "new tables" and not repeats of old ones.

As a consequence, since they are created in a different order, some of the table names may get changed. For example, commonly `FE12_C0_D01_Q083` and `FE12_C1_D10_Q083` may appear, where they would have both had "C0" before. But the values are as before.

By treating each subelement with its own "offset" and "stride", MixedElements will now be computed by block, rather than monolithically, so there may be a slight improvement to performance.

### CODE REVIEW

**Comment by michalhabera** [comment]
Some description?

**Review by garth-wells — APPROVED** [review]

File: ffcx/element_interface.py [hunk 1]
```diff
@@ -200,9 +179,32 @@ def reference_topology(self):
     def reference_geometry(self):
         return basix.geometry(self.element.cell_type)
 
-    @property
-    def is_blocked(self):
-        return False
+
+class ComponentElement(BaseElement):
+    def __init__(self, element, component):
```
Comment (garth-wells) [inline ffcx/element_interface.py:184]: Doc string and type hints?

File: ffcx/element_interface.py [hunk 2]
```diff
@@ -223,6 +225,23 @@ def tabulate(self, nderivs, points):
             tables.append(new_table)
         return tables
 
+    def get_component_element(self, flat_component):
```
Comment (garth-wells) [inline ffcx/element_interface.py:228]: Doc string and type hint.

File: ffcx/ir/elementtables.py [hunk 3]
```diff
@@ -407,13 +259,14 @@ def permute_quadrature_quadrilateral(points, reflections=0, rotations=0):
...
 
 
-def build_element_tables(quadrature_rule,
-                         cell,
-                         integral_type,
-                         entitytype,
-                         modified_terminals,
-                         rtol=default_rtol,
-                         atol=default_atol):
+def build_optimized_tables(quadrature_rule,
```
Comment (garth-wells) [inline ffcx/ir/elementtables.py:262]: Could type hints be added?

**Review by garth-wells — COMMENTED** [review]
Could some doc strings and type hints be added?

**Comment by mscroggs** [comment]
I'm just adding some (much needed) doc strings to `element_interface.py`

**Pull request merged**